[Spec]: Added initial specification of Searchsorted op (#26887)
openvinotoolkit/openvino: docs/articles_en/documentation/openvino-ir-format/operation-sets/operation-specs/sort/search-sorted-15.rst

SearchSorted
===============

.. meta::
  :description: Learn about SearchSorted - a sorting and maximization
                operation, which requires two input tensors.

**Versioned name**: *SearchSorted-15*

**Category**: *Sorting and maximization*

**Short description**: Find the indices from the innermost dimension of sorted_sequence such that, if the corresponding values in values were inserted before the indices, when sorted, the order of the corresponding innermost dimension within sorted_sequence would be preserved.

> **Detailed description**: `Reference <https://pytorch.org/docs/stable/generated/torch.searchsorted.html>`__.

Reviewer: Please give our normal description for this operation and share examples for this operation. Just give a reference to PyTorch is not good since it can be changing so our operation is also automatically changing.

Author: 1. What is a 'normal description for this operation'?

2. So I will remove the reference at all, since the only refs I have found are from pyTorch or Numpy, further more, I was explicitly told to do it exactly as in Pytorch, since fast path is needed foe this op.

Reviewer: Let us have our own description and give examples.
What if they change description in PyTorch we will have to re-implement our functionality.

Author: Spec is rewritten - hope will be fine now.

**Attributes**

> * *right*

Reviewer 2: This does not seem to render correctly in rich diff:
![image](https://github.com/user-attachments/assets/68d1515c-afd3-48e0-a9e6-9fe65217b872)

Author: Fixed.

>   * **Description**: *right*  if False, return the first suitable location that is found. If True, return the last such index. If no suitable index found, return 0 for non-numerical value (eg. nan, inf) or the size of innermost dimension within sorted_sequence (one pass the last index of the innermost dimension). In other words, if False, gets the lower bound index for each value in values on the corresponding innermost dimension of the sorted_sequence. If True, gets the upper bound index instead. Default value is False. side does the same and is preferred. It will error if side is set to “left” while this is True.

Reviewer: ```suggestion
  * **Description**: *right*  if False, return the first suitable location that is found. If True, return the last such index. If no suitable index is found, return 0 for non-numerical value (e.g. nan, inf) or the size of innermost dimension within sorted_sequence (one pass the last index of the innermost dimension). In other words, if False, get the lower bound index for each value in values on the corresponding innermost dimension of the sorted_sequence. If True, get the upper bound index instead. Default value is False. Side does the same and is preferred. It will error if side is set to “left” while this is True.
```

  * **Range of values**: true or false
  * **Type**: boolean
  * **Default value**: false
  * **Required**: *no*

**Inputs**:

* **1**: ND input tensor of type *T*, containing monotonically increasing sequence on the innermost dimension **Required.**

> * **2**: ND input tensor of type *T*, containing the search values. **Required.**

Author: Scalar is also a special case of nd tensor, isn't it?

> * **2**: ND input tensor of type *T*, containing the search values. **Required.**

Reviewer 3: In fact scalar can be interpreted as 0D case of ND. In my opinion the description for the second input is ok, but for the same reason, it should be mentioned that the first input (sorted sequence) can be ND but not scalar (0D). Also we need to add (maybe as a note above) that if sorted sequence is 1D, then the values can have any shape, otherwise the rank should be equal.

> * **2**: ND input tensor of type *T*, containing the search values. **Required.**

Author: Why 1st input cannot be a scalar then? It works with pytorch.

> * **2**: ND input tensor of type *T*, containing the search values. **Required.**

Reviewer 3: > Why 1st input cannot be a scalar then? It works with pytorch.

Based on my tests it doesn't. 
Cases like in0: scalar, in1: scalar or in0: scalar, in1: 1D are leading to exception that it's not allowed.
And preventing 0D sequence input is reasonable for me, as index of insertion to "scalar" shape could be misleading.
Could you share your example?

> * **2**: ND input tensor of type *T*, containing the search values. **Required.**

Author: (some discussion has happened via external channels) Fixed.

> * **2**: ND input tensor of type *T*, containing the search values. **Required.**

Reviewer 3: > Also we need to add (maybe as a note above) that if sorted sequence is 1D, then the values can have any shape, otherwise the rank should be equal.

Ok, what about this note?

> * **2**: ND input tensor of type *T*, containing the search values. **Required.**

Author: Fixed - hope it is fine now, otherwise please push a commit with whatever changes you also want to avoid any confusion please.

**Outputs**:

* **1**: Tensor of type *TOut*, with the same shape as second input tensor, containing the indices.

**Types**

* *T*: any supported floating-point and integer type.

* *TOut*: int64.

**Example**

.. code-block:: xml
   :force:

   <layer ... type="SearchSorted" ... >

>        <right="True"/>

Reviewer 2: ```suggestion
       <data right="True"/>
```

       <input>
           <port id="0">
               <dim>7</dim>
               <dim>256</dim>
               <dim>200</dim>
               <dim>200</dim>
           </port>
           <port id="1">
               <dim>7</dim>
               <dim>256</dim>
               <dim>200</dim>
               <dim>10</dim>
           </port>
       </input>
       <output>

>            <port id="0" precision="INT64">

Reviewer 2: ```suggestion
           <port id="0" precision="I64">
```
According to other files we're following this convention.

**Category**: *Sorting and maximization*

**Short description**: Determines the indices in the innermost dimension of a sorted sequence where elements should be inserted to maintain order.

**Detailed description**: *SearchSorted* operation determines the indices in the innermost dimension of a sorted sequence where elements should be inserted to maintain order. The operation is based on the binary search algorithm. The operation is performed on two input tensors: the first tensor contains a monotonically increasing sequence on the innermost dimension, and the second tensor contains the search values. The operation returns a tensor with the same shape as the second input tensor, containing the indices.

**Attributes**

* *right*

  * **Description**: If False, set the first suitable index. If True, return the last suitable index for given value. Default is False.
  * **Range of values**: true or false
  * **Type**: boolean
  * **Default value**: false
  * **Required**: *no*

**Inputs**:

* **1**: ``sorted`` - ND input tensor of type *T* - cannot be a scalar, containing monotonically increasing sequence on the innermost dimension. **Required.**

> * **2**: ``values`` - ND input tensor of type *T*, containing the search values. If sorted sequence is 1D, then the values can have any shape, otherwise the rank should be equal to sorted input. **Required.**

Reviewer 3: ```suggestion
* **2**: ``values`` - ND input tensor of type *T*, containing the search values. If sorted sequence is 1D, then the values can have any shape, otherwise the rank should be equal to the rank of sorted input. **Required.**
```
